src/System.Runtime.Extensions/src/System/Runtime/CompilerServices/SwitchExpressionException.cs

```diff
@@ -0,0 +1,65 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Serialization;
+
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>
+    /// Indicates that a switch expression that was non-exhaustive failed to match its input
+    /// at runtime, e.g. in the C# 8 expression <code>3 switch { 4 => 5 }</code>.
+    /// The exception optionally contains an object representing the unmatched value.
+    /// </summary>
+    [Serializable]
+    [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
+    public sealed class SwitchExpressionException : InvalidOperationException
+    {
+        public SwitchExpressionException()
+            : base(SR.Arg_SwitchExpressionException) { }
+
+        public SwitchExpressionException(object unmatchedValue)
+            : this()
+        {
+            UnmatchedValue = unmatchedValue;
+        }
+
+        private SwitchExpressionException(SerializationInfo info, StreamingContext context)
+            : base(info, context)
+        {
+            if (info == null)
+            {
+                throw new ArgumentNullException(nameof(info));
+            }
+            UnmatchedValue = info.GetValue(nameof(UnmatchedValue), typeof(object));
+        }
+
+        public object UnmatchedValue { get; }
+
+        public override void GetObjectData(SerializationInfo info, StreamingContext context)
+        {
+            base.GetObjectData(info, context);
+            if (info == null)
+            {
+                throw new ArgumentNullException(nameof(info));
+            }
+            info.AddValue(nameof(UnmatchedValue), UnmatchedValue, typeof(object));
+        }
+
+        public override string Message
+        {
+            get
+            {
+                string s = base.Message;
```

Comment:
This whole constructs look weird. Prefer something more readable:

```c#
if (UnmatchedValue is null) base.Message;

string valueMessage = SR.Format(SR.SwitchExpressionException_UnmatchedValue, UnmatchedValue.ToString());
string message = base.Message;

return s is null ? valueMessage : message + Environment.NewLine + valueMessage;
```

Reply:
I wasn't sure if I need to add a variable if it's going to only be used once in the code block. (e.g. message in your example above.)
For readability maybe I should add back valueMessage string though.